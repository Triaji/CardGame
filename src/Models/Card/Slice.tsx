import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '@store/Config';

export interface CardState {
  id: number;
  value: number;
  flipped: boolean;
  matched: boolean;
}

const initialState: CardState[] = [];

export const cardSlice = createSlice({
  name: 'cards',
  initialState,
  reducers: {
    addCard: (states, action: PayloadAction<CardState>) => {
      states.push(action.payload);
      return states;
    },
    flipCard: (states, action: PayloadAction<number>) =>
      states.map(state => {
        if (state.id !== action.payload) {
          return state;
        }

        return {
          ...state,
          flipped: true,
        };
      }),
    unflipCards: (states, action: PayloadAction<number[]>) =>
      states.map(state => {
        if (action.payload.indexOf(state.id) === -1) {
          return state;
        }

        return {
          ...state,
          flipped: false,
        };
      }),
    unflipCardsAll: states =>
      states.map(state => {
        return {
          ...state,
          flipped: false,
        };
      }),
    matchCards: (states, action: PayloadAction<number[]>) =>
      states.map(state => {
        if (action.payload.indexOf(state.id) === -1) {
          return state;
        }

        return {
          ...state,
          matched: true,
        };
      }),
    setCards: (states, action: PayloadAction<CardState[]>) => action.payload,
  },
});

export const {
  addCard,
  flipCard,
  unflipCards,
  unflipCardsAll,
  setCards,
  matchCards,
} = cardSlice.actions;

export const selectCards = (state: RootState) => state.cards;

export default cardSlice.reducer;
