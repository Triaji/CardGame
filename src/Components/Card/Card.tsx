import React, { useRef, useEffect, useCallback, useState } from 'react';
import {
  Animated,
  StyleSheet,
  Text,
  TouchableWithoutFeedback,
  Image,
} from 'react-native';

import { CardState } from '@models/Card/Slice';
import { cardBack } from '@assets/Images';

type Props = {
  card: CardState;
  reset?: Boolean;
  onPress: (value: CardState) => void;
};

const Card = (props: Props) => {
  const { card, onPress } = props;
  const [prevState, setPrevState] = useState<CardState>();
  const flip = useRef(new Animated.Value(0)).current;
  const shake = useRef(new Animated.Value(0)).current;

  const flipToFrontStyle = {
    transform: [
      { perspective: 1000 },
      {
        rotateY: flip.interpolate({
          inputRange: [0, 180],
          outputRange: ['0deg', '180deg'],
        }),
      },
    ],
  };

  const flipToBackStyle = {
    transform: [
      { perspective: 1000 },
      {
        rotateY: flip.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg'],
        }),
      },
    ],
  };

  const shakeStyle = {
    transform: [{ translateX: shake }],
  };

  const flipToFront = useCallback(() => {
    Animated.spring(flip, {
      toValue: 180,
      friction: 8,
      tension: 10,
      useNativeDriver: true,
    }).start();
  }, [flip]);

  const flipToBack = useCallback(() => {
    Animated.spring(flip, {
      toValue: 0,
      friction: 8,
      tension: 10,
      useNativeDriver: true,
    }).start();
  }, [flip]);

  const shakeAnimation = useCallback(() => {
    Animated.sequence([
      Animated.timing(shake, {
        toValue: 10,
        duration: 100,
        useNativeDriver: true,
      }),
      Animated.timing(shake, {
        toValue: -10,
        duration: 100,
        useNativeDriver: true,
      }),
      Animated.timing(shake, {
        toValue: 10,
        duration: 100,
        useNativeDriver: true,
      }),
      Animated.timing(shake, {
        toValue: 1,
        duration: 100,
        useNativeDriver: true,
      }),
    ]).start();
  }, [shake]);

  const flipCard = useCallback(() => {
    onPress && onPress(card);
  }, [onPress, card]);

  useEffect(() => {
    if (card.flipped) {
      flipToFront();
    } else {
      if (prevState?.flipped) {
        if (!prevState?.matched) {
          shakeAnimation();
        }
        setTimeout(
          () => {
            flipToBack();
          },
          !prevState?.matched ? 500 : 0,
        );
      }
    }

    setPrevState(card);
  }, [card, prevState, flipToFront, flipToBack, shakeAnimation]);

  return (
    <TouchableWithoutFeedback
      testID="touchable-card"
      style={[styles.container]}
      onPress={() => flipCard()}>
      <Animated.View style={[styles.cardContainer, { ...shakeStyle }]}>
        <Animated.View
          style={[styles.card, styles.cardFront, { ...flipToBackStyle }]}>
          <Text style={styles.cardText}>{card.value}</Text>
        </Animated.View>
        <Animated.View
          style={[styles.card, styles.cardBack, { ...flipToFrontStyle }]}>
          <Image
            source={cardBack}
            style={styles.cardBackImage}
            resizeMode="stretch"
          />
        </Animated.View>
      </Animated.View>
    </TouchableWithoutFeedback>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  cardContainer: { flex: 1, padding: 5 },
  card: {
    flex: 1,
    borderRadius: 8,
    borderWidth: 10,
    borderColor: '#FFF',
    alignItems: 'center',
    justifyContent: 'center',
    backfaceVisibility: 'hidden',
  },
  cardFront: {
    backgroundColor: '#FFCC00',
  },
  cardBack: {
    borderWidth: 5,
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 5,
    left: 5,
  },
  cardText: {
    fontFamily: 'Silkscreen-Regular',
    fontSize: 40,
  },
  cardBackImage: {
    width: '100%',
    height: '100%',
    backgroundColor: '#FFF',
  },
});

export default Card;
