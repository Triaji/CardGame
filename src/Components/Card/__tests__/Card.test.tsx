import React from 'react';
import { fireEvent, screen, render } from '@testing-library/react-native';
import Card from '../Card';
import { default as RN, Animated } from 'react-native';

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');
jest.useFakeTimers();

describe('CardComponent', () => {
  it('should call onPress correctly when card is pressed', () => {
    const mockPress = jest.fn();
    render(
      <Card
        card={{ id: 0, value: 10, flipped: false, matched: false }}
        onPress={mockPress}
      />,
    );

    fireEvent.press(screen.getByTestId('touchable-card'));
    expect(mockPress).toHaveBeenCalled();
  });

  it('should renders correctly', () => {
    const card = render(
      <Card
        card={{ id: 0, value: 10, flipped: false, matched: false }}
        onPress={() => {}}
      />,
    );
    expect(screen.getByText(/10/i)).toBeDefined();
    card.rerender(
      <Card
        card={{ id: 0, value: 10, flipped: true, matched: false }}
        onPress={() => {}}
      />,
    );

    card.rerender(
      <Card
        card={{ id: 0, value: 10, flipped: false, matched: false }}
        onPress={() => {}}
      />,
    );

    card.rerender(
      <Card
        card={{ id: 0, value: 10, flipped: true, matched: true }}
        onPress={() => {}}
      />,
    );

    card.rerender(
      <Card
        card={{ id: 0, value: 10, flipped: false, matched: false }}
        onPress={() => {}}
      />,
    );

    jest.advanceTimersByTime(1000);

    card.toJSON();

    expect(card).toMatchSnapshot();
  });
});
