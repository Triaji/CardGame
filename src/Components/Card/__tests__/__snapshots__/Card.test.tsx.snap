// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardComponent should renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "padding": 5,
      "transform": Array [
        Object {
          "translateX": 1,
        },
      ],
    }
  }
  testID="touchable-card"
>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#FFCC00",
        "borderColor": "#FFF",
        "borderRadius": 8,
        "borderWidth": 10,
        "flex": 1,
        "justifyContent": "center",
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateY": "180deg",
          },
        ],
      }
    }
  >
    <Text
      style={
        Object {
          "fontFamily": "Silkscreen-Regular",
          "fontSize": 40,
        }
      }
    >
      10
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "borderColor": "#FFF",
        "borderRadius": 8,
        "borderWidth": 5,
        "flex": 1,
        "height": "100%",
        "justifyContent": "center",
        "left": 5,
        "position": "absolute",
        "top": 5,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateY": "0deg",
          },
        ],
        "width": "100%",
      }
    }
  >
    <Image
      resizeMode="stretch"
      source={
        Object {
          "testUri": "../../../src/Assets/Images/card-back.png",
        }
      }
      style={
        Object {
          "backgroundColor": "#FFF",
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
